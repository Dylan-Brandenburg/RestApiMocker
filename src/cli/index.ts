import * as arg from 'arg'
import * as path from 'path'

import Restapify from '../server'

export const cli = (cliArgs: string[]): void => {
  const args = arg({
    // Types
    '--help': Boolean,
    '-h': '--help',

    '--port': Number,
    '-p': '--port',

    '--baseURL': String,
    '--b': '--baseURL',

    '--dir': String,
    '-d': '--dir'
  }, { argv: cliArgs })

  const {
    '--dir': rootDir = '.',
    '--port': port,
    '--baseURL': baseURL
  } = args

  const RestapifyInstance = new Restapify({
    rootDir: path.resolve(rootDir),
    port,
    baseURL
  })
}
