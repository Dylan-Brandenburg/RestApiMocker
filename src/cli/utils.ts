import * as chalk from 'chalk'
import * as boxen from 'boxen'

import { HttpVerb } from '../types'

export const getMethodOutput = (method: HttpVerb): string => {
  let methodOutput

  switch (method) {
  case 'DELETE':
    methodOutput = chalk.red
    break
  case 'POST':
    methodOutput = chalk.yellow
    break
  case 'PUT':
    methodOutput = chalk.blue
    break
  case 'PATCH':
    methodOutput = chalk.gray
    break

  default: case 'GET':
    methodOutput = chalk.green
    break
  }

  let methodName = method
  let methodNameLength = method.length

  for (let index = 0; index < (6 - methodNameLength); index += 1) {
    methodName += ' '
  }

  methodOutput = methodOutput.bold(`${methodName}`)

  return methodOutput
}

export const getInstanceOverviewOutput = (port: number, apiBaseURL: string): string => {
  const apiBaseURLLink = chalk.blue(`http://localhost:${port}${apiBaseURL}`)
  const dashboardURLLink = chalk.blue(`http://localhost:${port}/restapify`)
  const apiBaseURLOutput = `ðŸ“¦mocked API is served from ${apiBaseURLLink}`
  const dashboardURLOutput = `ðŸŽ› dashboard is running on ${dashboardURLLink}`
  return boxen(`${apiBaseURLOutput}\n${dashboardURLOutput} `, { padding: 1, borderColor: 'magenta' })
}
