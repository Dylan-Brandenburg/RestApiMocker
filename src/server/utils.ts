import { HTTP_VERBS } from './CONST'

export const getVarsInPath = (path: string): string[] => {
  const vars: string[] = []

  if (path.endsWith('.json')) {
    path = path.slice(0, -'.json'.length)
  }

  const explodedPath = path.split('/')

  explodedPath.forEach(pathElement => {
    const isVar = pathElement.startsWith('[') && pathElement.endsWith(']')
    if (isVar) {
      vars.push(pathElement.slice(1, -1))
    }
  })

  return vars
}

export const isHttpVerb = (str: string): boolean => {
  // @ts-ignore
  return HTTP_VERBS.includes(str)
}

export const isStateVariable = (str: string): boolean => {
  return str.startsWith('{') && str.endsWith('}')
}

export const isNumeric = (str:string):boolean=> {
  return !Number.isNaN(str) // use type coercion to parse the _entirety_ of the string (`parseFloat` alone does not do this)...
         && !Number.isNaN(parseFloat(str)) // ...and ensure strings of whitespace fail
}

export const routeResolve = (...routes: string[]): string => {
  let finalRoute = ''

  routes.forEach((route, routeId) => {
    const hasPreviousRouteFinalSlash = !!routes[routeId - 1]?.endsWith('/')
    const hasRouteFirstSlash = route.startsWith('/')

    if (hasPreviousRouteFinalSlash && hasRouteFirstSlash) {
      finalRoute += route.slice(1)
    } else if (!hasPreviousRouteFinalSlash && !hasRouteFirstSlash) {
      finalRoute += '/' + route
    } else {
      finalRoute += route
    }
  })

  return finalRoute
}

export const withoutUndefinedFromObject = (obj: Object): Object => {
  // @ts-ignore
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key])
  return obj
}
