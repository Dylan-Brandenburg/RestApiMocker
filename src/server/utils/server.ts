export const getVarsInPath = (path: string): string[] => {
  const vars: string[] = []

  if (path.endsWith('.json')) {
    path = path.slice(0, -'.json'.length)
  }

  const explodedPath = path.split('/')

  explodedPath.forEach(pathElement => {
    const isVar = pathElement.startsWith('[') && pathElement.endsWith(']')
    if (isVar) {
      vars.push(pathElement.slice(1, -1))
    }
  })

  return vars
}
