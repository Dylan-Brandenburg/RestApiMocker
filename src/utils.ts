export const replaceAll = (str: string, find: string, replace: string): string => {
  return str.split(find).join(replace)
}
