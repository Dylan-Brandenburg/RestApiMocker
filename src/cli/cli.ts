import * as arg from 'arg'
import * as path from 'path'

import Restapify from '../Restapify'
import { getInstanceOverviewOutput, getMethodOutput, onRestapifyInstanceError } from './utils'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const cli = ([nodePath, scriptPath, entryFolder, ...cliArgs]: string[]): void => {
  const args = arg({
    '--help': Boolean,
    '-h': '--help',

    '--port': Number,
    '-p': '--port',

    '--baseURL': String,
    '-b': '--baseURL',

    '--no-open': Boolean
  }, { argv: cliArgs })

  const {
    '--port': port,
    '--baseURL': baseURL,
    '--no-open': noOpen = false
  } = args

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const RestapifyInstance = new Restapify({
    rootDir: path.resolve(entryFolder),
    port,
    baseURL,
    openDashboard: !noOpen
  })
  RestapifyInstance.on('server:start', () => {
    console.log(`\n🏗 Try to serve on port ${RestapifyInstance.port}`)
  })
  RestapifyInstance.onError(({ error }) => {
    onRestapifyInstanceError(error, {
      rootDir: RestapifyInstance.rootDir,
      apiBaseUrl: RestapifyInstance.apiBaseUrl,
      port: RestapifyInstance.port
    })
  })

  RestapifyInstance.on('start', () => {
    const servedRoutes = RestapifyInstance.getServedRoutes()

    servedRoutes.forEach(servedRoute => {
      let methodOutput = getMethodOutput(servedRoute.method)

      console.log(`${methodOutput} ${servedRoute.route}`)
    })

    console.log(getInstanceOverviewOutput(
      RestapifyInstance.port,
      RestapifyInstance.apiBaseUrl
    ))
  })

  RestapifyInstance.run()
}
