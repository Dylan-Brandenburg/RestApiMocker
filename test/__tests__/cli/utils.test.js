import { getInstanceOverviewOutput } from '../../../src/cli/utils'

describe('CLI utils functions', () => {
  it('should output overview', () => {
    const port = 6767
    const baseUrl = 'dev'
    const output = getInstanceOverviewOutput(port, baseUrl)

    expect(output).toMatchSnapshot()
  })
})