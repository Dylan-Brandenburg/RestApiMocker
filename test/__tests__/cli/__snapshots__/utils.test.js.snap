// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI utils functions should output overview 1`] = `
"[35m┌────────────────────────────────────────────────────┐[39m
[35m│[39m                                                    [35m│[39m
[35m│[39m   [35m🚀 Restapify is running:[39m                         [35m│[39m
[35m│[39m                                                    [35m│[39m
[35m│[39m   [1m- 📦API entry point:[22m [94mhttp://localhost:6767/dev[39m   [35m│[39m
[35m│[39m   [1m- 🎛 Dashboard:[22m [94mhttp://localhost:6767/restapify[39m   [35m│[39m
[35m│[39m                                                    [35m│[39m
[35m│[39m   [93mUse Ctrl+C to quit this process[39m                  [35m│[39m
[35m│[39m                                                    [35m│[39m
[35m└────────────────────────────────────────────────────┘[39m"
`;
